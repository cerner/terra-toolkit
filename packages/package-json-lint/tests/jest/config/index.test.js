jest.mock('cosmiconfig');

const { cosmiconfig } = require('cosmiconfig');
const path = require('path');

const { getConfigForFile } = require('../../../src/config');

describe('config', () => {
  describe('getConfigForFile', () => {
    it('returns an empty object when there is no config defined for a project', async () => {
      const search = jest.fn();
      cosmiconfig.mockReturnValueOnce({ search });
      search.mockResolvedValueOnce(undefined);

      const config = await getConfigForFile({ packageJsonPath: '/x/y/z/terra-toolkit/package.json' });

      expect(config).toEqual({});

      expect(cosmiconfig).toHaveBeenCalledWith('package-json-lint');
      expect(search).toHaveBeenCalledWith('/x/y/z/terra-toolkit');
    });

    it('returns an extended config with no transitive extend', async () => {
      const search = jest.fn();
      cosmiconfig.mockReturnValueOnce({ search });
      const specifiedConfig = { extends: 'package-json-lint-config-mock', rules: { x: 1, y: 2, z: 3 } };
      search.mockResolvedValueOnce({ config: specifiedConfig, filepath: path.join(__dirname, '..', '__mocks__', '/package.json') });

      const config = await getConfigForFile({ packageJsonPath: '/x/y/z/terra-toolkit/package.json' });

      expect(config).toMatchSnapshot();

      expect(cosmiconfig).toHaveBeenCalledWith('package-json-lint');
      expect(search).toHaveBeenCalledWith('/x/y/z/terra-toolkit');
    });

    it('returns an extended config with 3 transitive extends', async () => {
      const search = jest.fn();
      cosmiconfig.mockReturnValueOnce({ search });
      const specifiedConfig = { extends: 'package-json-lint-config-mock1', rules: { override: 100 } };
      search.mockResolvedValueOnce({ config: specifiedConfig, filepath: path.join(__dirname, '..', '__mocks__', '/package.json') });

      const config = await getConfigForFile({ packageJsonPath: '/x/y/z/terra-toolkit/package.json' });

      expect(config).toMatchSnapshot();

      expect(cosmiconfig).toHaveBeenCalledWith('package-json-lint');
      expect(search).toHaveBeenCalledWith('/x/y/z/terra-toolkit');
    });

    it('returns an extended config with multiple extenions', async () => {
      const search = jest.fn();
      cosmiconfig.mockReturnValueOnce({ search });
      const specifiedConfig = {
        extends: [
          'package-json-lint-config-mock3',
          'package-json-lint-config-mock2',
          'package-json-lint-config-mock1',
          'package-json-lint-config-mock',
        ],
        rules: { override: 100 },
      };
      search.mockResolvedValueOnce({ config: specifiedConfig, filepath: path.join(__dirname, '..', '__mocks__', '/package.json') });

      const config = await getConfigForFile({ packageJsonPath: '/x/y/z/terra-toolkit/package.json' });

      expect(config).toMatchSnapshot();

      expect(cosmiconfig).toHaveBeenCalledWith('package-json-lint');
      expect(search).toHaveBeenCalledWith('/x/y/z/terra-toolkit');
    });

    it('returns an extended config with multiple extenions and no override', async () => {
      const search = jest.fn();
      cosmiconfig.mockReturnValueOnce({ search });
      const specifiedConfig = {
        extends: [
          'package-json-lint-config-mock',
          'package-json-lint-config-mock1',
          'package-json-lint-config-mock2',
          'package-json-lint-config-mock3',
        ],
      };
      search.mockResolvedValueOnce({ config: specifiedConfig, filepath: path.join(__dirname, '..', '__mocks__', '/package.json') });

      const config = await getConfigForFile({ packageJsonPath: '/x/y/z/terra-toolkit/package.json' });

      expect(config).toMatchSnapshot();

      expect(cosmiconfig).toHaveBeenCalledWith('package-json-lint');
      expect(search).toHaveBeenCalledWith('/x/y/z/terra-toolkit');
    });
  });
});
