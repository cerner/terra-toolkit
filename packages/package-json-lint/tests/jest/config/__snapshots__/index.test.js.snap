// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config getConfigForFile returns an extended config with 3 transitive extends 1`] = `
Object {
  "rules": Object {
    "a": 1,
    "b": 2,
    "c": 3,
    "override": 100,
  },
}
`;

exports[`config getConfigForFile returns an extended config with multiple extenions 1`] = `
Object {
  "rules": Object {
    "a": 1,
    "b": 2,
    "c": 3,
    "override": 100,
    "x": 4,
    "y": 5,
    "z": 6,
  },
}
`;

exports[`config getConfigForFile returns an extended config with multiple extenions and no override 1`] = `
Object {
  "rules": Object {
    "a": 1,
    "b": 2,
    "c": 3,
    "override": 3,
    "x": 4,
    "y": 5,
    "z": 6,
  },
}
`;

exports[`config getConfigForFile returns an extended config with no transitive extend 1`] = `
Object {
  "rules": Object {
    "a": 1,
    "b": 2,
    "c": 3,
    "x": 1,
    "y": 2,
    "z": 3,
  },
}
`;
